File: apple/HavenApp/ReportUI.swift
import SwiftUI

// Decentralized moderation (App Review 1.2). Haven circles have no owner and the developer holds
// no keys, so moderation is the members': a report is sealed to the WHOLE circle (like a
// SensitiveFlag) and every member acts with the power they already hold — hide for themselves,
// remove the author from their circle, or block. The only thing that ever leaves the circle is a
// content-free ledger entry: identity vs identity, action, offense category. No PII, no content.

/// Fire-and-forget, content-free entries to the developer ledger on the push Worker. Node ids are
/// opaque public keys — the ledger records only that an identity WAS REPORTED and for which
/// category, never what was said or shown. This is the 1.2 "notify the developer" hook.
///
/// Only an explicit report comes here (audit F1). **Blocking never touches the network**: it is a
/// private, local decision to stop seeing someone, and it stays on the device.
enum ModerationLedger {
    @MainActor
    static func report(subject: String, reason: String) {
        guard !subject.isEmpty, let url = URL(string: PushManager.relay + "/flag") else { return }
        // Signed with the identity key (audit F1): the Terms attach real consequences to a ledger row,
        // so an unsigned POST must not be able to plant one. The signature binds subject + action +
        // category, so a captured flag can't be re-aimed at someone else. Unsigned = we don't send.
        guard let seed = AccountStore.storedSeed(), let acct = try? Account.fromSeed(seed: seed) else { return }
        let category = String(reason.prefix(64))
        let ts = UInt64(Date().timeIntervalSince1970)
        let sig = acct.signPushRegistration(token: "flag-v1:\(subject):report:\(category)", tsSecs: ts)
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "actor": acct.nodeIdHex(),
            "subject": subject,
            "action": "report",
            "reason": category,
            "ts": ts,
            "sig": sig.base64EncodedString(),
        ]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        URLSession.shared.dataTask(with: req).resume()   // fire and forget — never blocks moderation
    }
}

/// The offense categories a reporter picks from. The category travels BOTH ways: sealed to the
/// circle (so members can judge) and to the ledger (so patterns are visible) — the free-text
/// comment goes ONLY to the circle.
enum ReportReason: String, CaseIterable, Identifiable {
    case harassment = "Harassment or bullying"
    case sexual = "Nudity or sexual content"
    case violence = "Violence or dangerous acts"
    case spam = "Spam or scam"
    case other = "Something else"
    var id: String { rawValue }
    var icon: String {
        switch self {
        case .harassment: return "exclamationmark.bubble"
        case .sexual: return "eye.slash"
        case .violence: return "exclamationmark.triangle"
        case .spam: return "envelope.badge.shield.half.filled"
        case .other: return "flag"
        }
    }
}

/// Report a post/message: pick a category, optionally add a note for the circle, optionally block
/// the author in the same motion. Submitting hides the post for the reporter instantly.
///
/// Layout is a hand-rolled column, NOT a Form: macOS Form renders labeled rows as a two-column
/// grid, so the note field's label pushed the whole sheet's content rightward, and every control
/// hugged the section insets. Each choice is a full-width glass pill; the field and toggle carry
/// exactly one glass surface each (no doubled backgrounds or borders — house rule).
struct ReportSheet: View {
    let item: FeedItemFfi
    let authorName: String
    @ObservedObject private var feed = FeedStore.shared
    @Environment(\.dismiss) private var dismiss
    @State private var reason: ReportReason?
    @State private var comment = ""
    @State private var alsoBlock = false

    var body: some View {
        #if os(macOS)
        HavenMacSheet("Report post") {
            formColumn
        } footer: {
            Button("Report") { submit() }
                .buttonStyle(BrandButtonStyle())
                .disabled(reason == nil)
                .opacity(reason == nil ? 0.5 : 1)
                .keyboardShortcut(.defaultAction)
        }
        #else
        NavigationStack {
            ZStack {
                HavenBackground()
                ScrollView { formColumn.padding(20) }
            }
            .navigationTitle("Report post")
            .havenInlineNavTitle()
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Report") { submit() }
                        .disabled(reason == nil)
                        .tint(HavenTheme.pink)
                }
            }
        }
        #endif
    }

    private var formColumn: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("What's wrong with it?")
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(.secondary)
                .padding(.leading, 4)
            ForEach(ReportReason.allCases) { r in
                Button {
                    withAnimation(HavenTheme.snappy) { reason = r }
                } label: {
                    HStack(spacing: 10) {
                        Image(systemName: r.icon)
                            .frame(width: 22)
                            .foregroundStyle(reason == r ? HavenTheme.pink : .secondary)
                        Text(r.rawValue).foregroundStyle(.primary)
                        Spacer()
                        if reason == r {
                            Image(systemName: "checkmark.circle.fill").foregroundStyle(HavenTheme.pink)
                        }
                    }
                    .padding(.horizontal, 16).padding(.vertical, 12)
                    .havenGlass(in: Capsule(), tint: reason == r ? HavenTheme.pink.opacity(0.35) : nil)
                    .contentShape(Capsule())
                }
                .buttonStyle(PressableStyle())
            }

            TextField("Add a note for your circle (optional)", text: $comment, axis: .vertical)
                .lineLimit(2 ... 4)
                .textFieldStyle(.plain)
                .padding(.horizontal, 16).padding(.vertical, 12)
                .havenGlass(in: RoundedRectangle(cornerRadius: 20, style: .continuous))
                .padding(.top, 8)

            // Label leading, switch trailing — a bare Toggle keeps them glued together and
            // centered, which looked lost inside the full-width pill.
            HStack {
                Label("Also block \(authorName)", systemImage: "hand.raised")
                Spacer()
                Toggle("", isOn: $alsoBlock)
                    .labelsHidden()
                    #if os(macOS)
                    .toggleStyle(.switch)   // a switch reads modern in the pill; the checkbox looked dated
                    #endif
                    .tint(HavenTheme.pink)
            }
            .padding(.horizontal, 16).padding(.vertical, 10)
            .havenGlass(in: Capsule())

            Text("The post disappears from your feed right away, and everyone in the circle sees your report so they can act too. Only who reported whom and the category are logged — never the content.")
                .font(.footnote)
                .foregroundStyle(.secondary)
                .padding(.horizontal, 4).padding(.top, 6)
        }
    }

    private func submit() {
        guard let reason else { return }
        let author = feed.report(circleId: feed.activeCircleId, target: item.id,
                                 reason: reason.rawValue, comment: comment)
        if alsoBlock, let author { feed.blockConnection(author) }
        dismiss()
    }
}

/// Shown on a post that OTHER members reported — the circle's shared moderation signal. Each
/// viewer decides for themselves: hide it, remove the author from their circle, or block. The
/// reporter never sees this (the post is already hidden for them).
struct ReportedBanner: View {
    let item: FeedItemFfi
    let authorName: String
    let reports: [ReportFfi]
    @ObservedObject private var feed = FeedStore.shared
    @State private var confirmRemove = false

    private var reporterNames: String {
        let names = Set(reports.map {
            ContactsStore.shared.name(forNodePrefix: $0.reporterShort) ?? "A member"
        })
        return names.sorted().joined(separator: ", ")
    }

    var body: some View {
        HStack(spacing: 8) {
            Image(systemName: "flag.fill").foregroundStyle(.orange)
            VStack(alignment: .leading, spacing: 1) {
                Text("Reported by \(reporterNames)").font(.caption.weight(.semibold))
                Text(reports.map(\.reason).uniqued().joined(separator: " · "))
                    .font(.caption2).foregroundStyle(.secondary)
            }
            Spacer()
            Menu {
                Button {
                    HiddenStore.shared.hide(item.id)
                    feed.refresh()
                } label: { Label("Hide for me", systemImage: "eye.slash") }
                Button(role: .destructive) { confirmRemove = true } label: {
                    Label("Remove \(authorName) from circle", systemImage: "person.badge.minus")
                }
                Button(role: .destructive) {
                    if let author = reports.first?.author { feed.blockConnection(author) }
                } label: { Label("Block \(authorName)", systemImage: "hand.raised.fill") }
            } label: {
                Text("Act").font(.caption.weight(.semibold)).foregroundStyle(HavenTheme.pink)
            }
            // Borderless or macOS wraps the bare label in a popup-button bezel.
            #if os(macOS)
            .menuStyle(.borderlessButton)
            .fixedSize()
            #endif
        }
        .padding(10)
        .background(.orange.opacity(0.12), in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        .confirmationDialog("Remove \(authorName) from this circle?", isPresented: $confirmRemove, titleVisibility: .visible) {
            Button("Remove", role: .destructive) {
                if let author = reports.first?.author {
                    feed.removeFromCircle(author, circleId: feed.activeCircleId)
                }
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("Their posts leave your view of the circle and they can't rejoin through you. Your own devices stay in sync.")
        }
    }
}

private extension Array where Element == String {
    /// Order-preserving dedup for the banner's reason line.
    func uniqued() -> [String] {
        var seen = Set<String>()
        return filter { seen.insert($0).inserted }
    }
}

File: android/app/src/main/java/com/blaineam/haven/ui/ReportUI.kt
package com.blaineam.haven.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import com.blaineam.haven.R
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.blaineam.haven.core.HavenNet
import com.blaineam.haven.core.HiddenStore
import com.blaineam.haven.core.REPORT_REASONS
import uniffi.haven_ffi.FeedItemFfi
import uniffi.haven_ffi.ReportFfi

// Decentralized moderation (parity with apple/HavenApp/ReportUI.swift — see docs/MODERATION.md).
// Haven circles have no owner and the developer holds no keys, so moderation is the members': a
// report is sealed to the WHOLE circle and every member acts with the power they already hold —
// hide for themselves, remove the author from their circle, or block. The only thing that ever
// leaves the circle is a content-free ledger entry: identity vs identity, action, category.

/** Report a post/message: pick a category, optionally add a note for the circle, optionally block
 *  the author in the same motion. Submitting hides the post for the reporter instantly. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReportSheet(item: FeedItemFfi, circleId: String, authorName: String, onDismiss: () -> Unit) {
    var reason by remember { mutableStateOf<String?>(null) }
    var comment by remember { mutableStateOf("") }
    var alsoBlock by remember { mutableStateOf(false) }
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = HavenTheme.card,
    ) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(bottom = 28.dp)) {
            Text(stringResource(R.string.report_title), color = HavenTheme.textPrimary, fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(4.dp))
            Text(stringResource(R.string.report_whats_wrong), color = HavenTheme.textSecondary, fontSize = 13.sp)
            Spacer(Modifier.height(10.dp))
            REPORT_REASONS.forEach { r ->
                Row(
                    Modifier.fillMaxWidth().padding(vertical = 3.dp).clip(RoundedCornerShape(12.dp))
                        .background(if (reason == r) HavenTheme.pink.copy(alpha = 0.18f) else HavenTheme.background)
                        .clickable { reason = r }
                        .padding(horizontal = 14.dp, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(r, color = HavenTheme.textPrimary, fontSize = 14.sp, modifier = Modifier.weight(1f))
                    if (reason == r) Icon(Icons.Filled.CheckCircle, null, tint = HavenTheme.pink, modifier = Modifier.size(18.dp))
                }
            }
            Spacer(Modifier.height(10.dp))
            OutlinedTextField(
                value = comment, onValueChange = { comment = it },
                placeholder = { Text(stringResource(R.string.report_note_placeholder), fontSize = 13.sp) },
                modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(14.dp), maxLines = 3,
                colors = OutlinedTextFieldDefaults.colors(focusedBorderColor = HavenTheme.pink, cursorColor = HavenTheme.pink),
            )
            Spacer(Modifier.height(6.dp))
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(stringResource(R.string.report_also_block, authorName), color = HavenTheme.textPrimary, fontSize = 14.sp, modifier = Modifier.weight(1f))
                Switch(
                    checked = alsoBlock, onCheckedChange = { alsoBlock = it },
                    colors = SwitchDefaults.colors(checkedTrackColor = HavenTheme.pink),
                )
            }
            Spacer(Modifier.height(6.dp))
            Text(
                stringResource(R.string.report_hides_note),
                color = HavenTheme.textSecondary, fontSize = 12.sp,
            )
            Spacer(Modifier.height(14.dp))
            BrandButton(text = stringResource(R.string.report_submit), modifier = Modifier.fillMaxWidth(), enabled = reason != null) {
                val author = HavenNet.report(circleId, item.id, reason ?: return@BrandButton, comment.trim())
                if (alsoBlock && author != null) HavenNet.block(author)
                onDismiss()
            }
        }
    }
}

/** Shown on a post that OTHER members reported — the circle's shared moderation signal. Each
 *  viewer decides for themselves: hide it, remove the author from their circle, or block. The
 *  reporter never sees this (the post is already hidden for them). */
@Composable
fun ReportedBanner(item: FeedItemFfi, circleId: String, authorName: String, reports: List<ReportFfi>) {
    var actMenu by remember(item.id) { mutableStateOf(false) }
    var confirmRemove by remember(item.id) { mutableStateOf(false) }
    val youLabel = stringResource(R.string.report_you)
    val reporterNames = reports.map { r ->
        if (r.reporter.startsWith(HavenNet.nodeIdHex.take(8))) youLabel else HavenNet.displayName(r.reporterShort)
    }.toSortedSet().joinToString(", ")
    val reasons = reports.map { it.reason }.distinct().joinToString(" · ")

    if (confirmRemove) {
        AlertDialog(
            onDismissRequest = { confirmRemove = false }, containerColor = HavenTheme.card,
            title = { Text(stringResource(R.string.report_remove_confirm_title, authorName), color = HavenTheme.textPrimary, fontSize = 16.sp) },
            text = {
                Text(stringResource(R.string.report_remove_confirm_body),
                    color = HavenTheme.textSecondary, fontSize = 13.sp)
            },
            confirmButton = {
                TextButton(onClick = {
                    confirmRemove = false
                    reports.firstOrNull()?.author?.let { HavenNet.removeFromCircle(circleId, it) }
                }) { Text(stringResource(R.string.common_remove), color = Color(0xFFF87171)) }
            },
            dismissButton = { TextButton(onClick = { confirmRemove = false }) { Text(stringResource(R.string.common_cancel), color = HavenTheme.textSecondary) } },
        )
    }

    Row(
        Modifier.fillMaxWidth().clip(RoundedCornerShape(12.dp))
            .background(Color(0xFFF59E0B).copy(alpha = 0.12f))
            .padding(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(Icons.Filled.Flag, null, tint = Color(0xFFF59E0B), modifier = Modifier.size(16.dp))
        Spacer(Modifier.width(8.dp))
        Column(Modifier.weight(1f)) {
            Text(stringResource(R.string.report_reported_by, reporterNames), color = HavenTheme.textPrimary, fontSize = 12.sp, fontWeight = FontWeight.SemiBold)
            Text(reasons, color = HavenTheme.textSecondary, fontSize = 11.sp)
        }
        Box {
            Text(stringResource(R.string.report_act), color = HavenTheme.pink, fontSize = 13.sp, fontWeight = FontWeight.SemiBold,
                modifier = Modifier.clip(RoundedCornerShape(8.dp)).clickable { actMenu = true }.padding(6.dp))
            DropdownMenu(expanded = actMenu, onDismissRequest = { actMenu = false },
                modifier = Modifier.background(HavenTheme.card)) {
                DropdownMenuItem(
                    text = { Text(stringResource(R.string.report_hide_for_me), color = HavenTheme.textPrimary) },
                    onClick = { actMenu = false; HiddenStore.hide(item.id) },
                )
                DropdownMenuItem(
                    text = { Text(stringResource(R.string.report_remove_from_circle, authorName), color = Color(0xFFF87171)) },
                    onClick = { actMenu = false; confirmRemove = true },
                )
                DropdownMenuItem(
                    text = { Text(stringResource(R.string.report_block_author, authorName), color = Color(0xFFF87171)) },
                    onClick = {
                        actMenu = false
                        reports.firstOrNull()?.author?.let { HavenNet.block(it) }
                    },
                )
            }
        }
    }
}
